src/System.Private.CoreLib/shared/System/Text/Rune.cs

```diff
@@ -0,0 +1,724 @@
+        /// </summary>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// If <paramref name="ch"/> represents a UTF-16 surrogate code point
+        /// U+D800..U+DFFF, inclusive.
+        /// </exception>
+        public Rune(char ch)
+        {
+            uint expanded = ch;
+            if (UnicodeUtility.IsSurrogateCodePoint(expanded))
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.ch);
+            }
+            ValueUnsigned = expanded;
+        }
+
+        /// <summary>
+        /// Creates a <see cref="Rune"/> from the provided Unicode scalar value.
+        /// </summary>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// If <paramref name="scalarValue"/> does not represent a value Unicode scalar value.
+        /// </exception>
+        public Rune(int value)
+            : this((uint)value)
+        {
+        }
+
+        /// <summary>
+        /// Creates a <see cref="Rune"/> from the provided Unicode scalar value.
+        /// </summary>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// If <paramref name="scalarValue"/> does not represent a value Unicode scalar value.
+        /// </exception>
+        [CLSCompliant(false)]
+        public Rune(uint value)
+        {
+            if (!UnicodeUtility.IsValidUnicodeScalar(value))
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.value);
+            }
+            ValueUnsigned = value;
+        }
+
+        // non-validating ctor
+        private Rune(uint scalarValue, bool unused)
+        {
+            UnicodeDebug.AssertIsValidScalar(scalarValue);
+            ValueUnsigned = scalarValue;
+        }
+
+        public static bool operator ==(Rune left, Rune right) => (left.ValueUnsigned == right.ValueUnsigned);
+
+        public static bool operator !=(Rune left, Rune right) => (left.ValueUnsigned != right.ValueUnsigned);
+
+        public static bool operator <(Rune left, Rune right) => (left.ValueUnsigned < right.ValueUnsigned);
+
+        public static bool operator <=(Rune left, Rune right) => (left.ValueUnsigned <= right.ValueUnsigned);
+
+        public static bool operator >(Rune left, Rune right) => (left.ValueUnsigned > right.ValueUnsigned);
+
+        public static bool operator >=(Rune left, Rune right) => (left.ValueUnsigned >= right.ValueUnsigned);
+
+        // Operators below are explicit because they may throw.
+
+        public static explicit operator Rune(char ch) => new Rune(ch);
+
+        [CLSCompliant(false)]
+        public static explicit operator Rune(uint value) => new Rune(value);
+
+        public static explicit operator Rune(int value) => new Rune(value);
+
+        // Displayed as "'<char>' (U+XXXX)"; e.g., "'e' (U+0065)"
+        private string DebuggerDisplay => FormattableString.Invariant($"'{(IsValid(ValueUnsigned) ? ToString() : "\uFFFD")}' (U+{ValueUnsigned:X4})");
```

Comment:
Do CR, LF, tab, NBSP, ZWJ, DEL, BS, and the like all print sensibly here?

Reply:
Mostly. I'll swap the `U+xxxx` and the character around so that the hex value always prints first, even if the character is non-printable or otherwise makes the display look strange.